src/backend/distributed/operations/shard_rebalancer.c

```diff
@@ -429,6 +434,123 @@ GetShardCost(uint64 shardId, void *voidContext)
+ */
+Datum
+worker_partitioned_relation_total_size(PG_FUNCTION_ARGS)
+{
+	char *relname = text_to_cstring(PG_GETARG_TEXT_P(0));
+
+	/* pg_partition_tree finds the partitions recursively and returns them */
+	char *queryText = "SELECT sum(pg_total_relation_size(relid))::bigint AS total_size "
+					  "FROM (SELECT relid from pg_partition_tree($1)) partition_tree;";
+
+	uint64 size = ExecSizeQueryWithRelName(queryText, relname);
+
+	PG_RETURN_UINT64(size);
+}
+
+
+/*
+ * worker_partitioned_relation_size returns the size (excluding indexes) of
+ * a partitioned relation with its partitions, using the function pg_partition_tree.
+ */
+Datum
+worker_partitioned_relation_size(PG_FUNCTION_ARGS)
+{
+	char *relname = text_to_cstring(PG_GETARG_TEXT_P(0));
+
+	/* pg_partition_tree finds the partitions recursively and returns them */
+	char *queryText = "SELECT sum(pg_relation_size(relid))::bigint AS total_size "
+					  "FROM (SELECT relid from pg_partition_tree($1)) partition_tree;";
+
+	uint64 size = ExecSizeQueryWithRelName(queryText, relname);
+
+	PG_RETURN_UINT64(size);
+}
+
+
+/*
+ * worker_partitioned_table_size returns the size (excluding indexes) of
+ * a partitioned table with its partitions, using the function pg_partition_tree.
+ * Unlike worker_partitioned_relation_size, this function includes the sizes of
+ * TOAST, vm and fsm for the given table. For further details, please see:
+ * https://www.postgresql.org/docs/13/functions-admin.html#FUNCTIONS-ADMIN-DBSIZE
+ */
+Datum
+worker_partitioned_table_size(PG_FUNCTION_ARGS)
+{
+	char *relname = text_to_cstring(PG_GETARG_TEXT_P(0));
+
+	/* pg_partition_tree finds the partitions recursively and returns them */
+	char *queryText = "SELECT sum(pg_table_size(relid))::bigint AS total_size "
+					  "FROM (SELECT relid from pg_partition_tree($1)) partition_tree;";
+
+	uint64 size = ExecSizeQueryWithRelName(queryText, relname);
+
+	PG_RETURN_UINT64(size);
+}
+
+
+/*
+ * ExecSizeQueryWithRelName is a helper function that executes the given size query
+ * using SPI, for the given relname parameter. Returns the size result.
+ * Errors out if any connection problem occurs with SPI.
+ */
+static uint64
+ExecSizeQueryWithRelName(char *sizeQuery, char *relname)
+{
+	uint64 size = 0;
+	const int paramCount = 1;
+	Oid paramTypes[1] = { TEXTOID };
+	Datum paramValues[1] = { CStringGetTextDatum(relname) };
+
+	int spiResult = SPI_connect();
+	if (spiResult != SPI_OK_CONNECT)
+	{
+		ereport(ERROR, (errmsg("could not connect to SPI manager"),
+						errdetail("couldn't get the size of partitioned shard named: %s",
+								  relname)));
+	}
+
+	spiResult = SPI_execute_with_args(sizeQuery, paramCount, paramTypes,
+									  paramValues,
+									  NULL, false, 0);
+
+	if (spiResult == SPI_OK_SELECT)
+	{
+		Assert(SPI_processed <= 1);
+		if (SPI_processed == 1)
+		{
+			bool isnull = false;
+
+			size = DatumGetInt64(SPI_getbinval(SPI_tuptable->vals[0],
```

Comment:
we are reading an `Int64` and assign it to an unsigned int. What is the reason we change the type here?